### PR TITLE
Bump elmPackages 

### PR DESCRIPTION
**Pull request by domenkozar**
Note that elm-verify-examples is still broken.

Fixes #320127 #267454 #304198

### CODE REVIEW

**Comment by domenkozar** [comment]
I pushed a fix for elm-pages.

**Review by PedroHLC — APPROVED** [review]
Diff LGTM

**Comment by domenkozar** [comment]
Seems to be https://github.com/NixOS/nixpkgs/issues/294183

File: pkgs/development/compilers/elm/packages/node/elm-pages/default.nix [hunk 1]
```diff
@@ -1,9 +1,29 @@
...
+        });
+      }
+    );
+in nodePkgs."elm-pages".overrideAttrs (
   old: {
+    inherit ESBUILD_BINARY_PATH;
     nativeBuildInputs = (old.nativeBuildInputs or [ ]) ++ [ makeWrapper old.nodejs.pkgs.node-gyp-build ];
 
+    preRebuild = ''
+      sed -i 's/"esbuild": "0\.19\.12"/"esbuild": "0.21.5"/' package.json
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/elm/packages/node/elm-pages/default.nix:24]: This should use substituteInPlace with --replace-fail

File: pkgs/development/compilers/elm/packages/node/generate-node-packages.sh [hunk 2]
```diff
@@ -10,3 +10,5 @@ $(nix-build $ROOT -A nodePackages.node2nix --no-out-link)/bin/node2nix \
     -o node-packages.nix \
     -c node-composition.nix \
     --no-copy-node-env -e ../../../../node-packages/node-env.nix
+# well, elm-pages requires two different version of esbuild so we twist it's wrist to only use one
+sed -i 's/sources."esbuild-0.19.12"/sources."esbuild-0.21.5"/' node-packages.nix
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/elm/packages/node/generate-node-packages.sh:14]: same here

File: pkgs/development/compilers/elm/packages/node/node-composition.nix [hunk 3]
```diff
@@ -8,7 +8,7 @@ let
   nodeEnv = import ../../../../node-packages/node-env.nix {
     inherit (pkgs) stdenv lib python2 runCommand writeTextFile writeShellScript;
     inherit pkgs nodejs;
-    libtool = if pkgs.stdenv.hostPlatform.isDarwin then pkgs.cctools or pkgs.darwin.cctools else null;
+    libtool = if pkgs.stdenv.isDarwin then pkgs.cctools or pkgs.darwin.cctools else null;
```
Comment (SuperSandro2000) [inline pkgs/development/compilers/elm/packages/node/node-composition.nix:11]: Why was this changed back?

**Review by SuperSandro2000 — COMMENTED** [review]
elm-packages should also eventually stop using node2nix but that's something for another PR

File: pkgs/development/compilers/elm/packages/node/node-composition.nix (hunk 3 above)
Comment (zupo) [reply pkgs/development/compilers/elm/packages/node/node-composition.nix:11]: Maybe because we used my mac to run the generation script. Could probably be reverted?

File: pkgs/development/compilers/elm/packages/node/elm-pages/default.nix [hunk 4]
```diff
@@ -1,9 +1,29 @@
 { nodePkgs, pkgs, lib, makeWrapper }:
 
-nodePkgs."elm-pages".overrideAttrs (
+let
+  ESBUILD_BINARY_PATH = lib.getExe (
+      pkgs.esbuild.override {
+        buildGoModule = args: pkgs.buildGoModule (args // rec {
+          version = "0.21.5";
```
Comment (turboMaCk) [inline pkgs/development/compilers/elm/packages/node/elm-pages/default.nix:7]: so I see we even need 2 different version of esbuild. I guess there was an issue with version of ES build when you tried to share the same one?

File: pkgs/development/compilers/elm/packages/node/default.nix [hunk 5]
```diff
@@ -6,6 +6,20 @@ let
...
+          src = pkgs.fetchFromGitHub {
+            owner = "evanw";
+            repo = "esbuild";
+            rev = "v${version}";
+            hash = "sha256-h/Vqwax4B4nehRP9TaYbdixAZdb1hx373dNxNHvDrtY=";
+          };
+          vendorHash = "sha256-+BfxCyg0KkDQpHt/wycy/8CTG6YBA/VJvJFhhzUnSiQ=";
+        });
+      }
+    );
```
[the remark this answers is not in the document]
Comment (zupo) [reply pkgs/development/compilers/elm/packages/node/default.nix:22]: It seemed very version sensitive when we tried. Had to provide the exact version.

[the remark this answers is not in the document]
Comment (turboMaCk) [reply pkgs/development/compilers/elm/packages/node/default.nix:22]: uh that will be a bit painful to maintain. Maybe it would make sense to port these away from node2nix to `buildNpmPackage` right away to prevent potential regression with future bulk updates.

File: pkgs/development/compilers/elm/packages/node/node-composition.nix (hunk 3 above)
Comment (zupo) [reply pkgs/development/compilers/elm/packages/node/node-composition.nix:11]: Reverted.

File: pkgs/development/compilers/elm/packages/node/elm-pages/default.nix (hunk 1 above)
Comment (zupo) [reply pkgs/development/compilers/elm/packages/node/elm-pages/default.nix:24]: Done.

File: pkgs/development/compilers/elm/packages/node/generate-node-packages.sh (hunk 2 above)
Comment (zupo) [reply pkgs/development/compilers/elm/packages/node/generate-node-packages.sh:14]: Not sure how to do this, since this is a shell script and it doesn't have `substituteInPlace`.

File: pkgs/development/compilers/elm/packages/node/default.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (zupo) [reply pkgs/development/compilers/elm/packages/node/default.nix:22]: Agree on it being painful. A slight condolence is the fact that Elm packages are usually quite slow to release new updates. 

I tried porting to `buildNpmPackage` but got lost. I'd need some hand-holding to get it done.

File: pkgs/development/compilers/elm/packages/node/elm-pages/default.nix (hunk 4 above)
Comment (zupo) [reply pkgs/development/compilers/elm/packages/node/elm-pages/default.nix:7]: correct

**Review by turboMaCk — APPROVED** [review]
LGTM

**Comment by domenkozar** [comment]
Note that the binary still fails to run because we need to also patch up `rollup`:

```
❯ /nix/store/mbbvs7bg3n90gj0fva1zxx5fp70fc9lq-elm-land-0.20.1/bin/elm-land 
/nix/store/mbbvs7bg3n90gj0fva1zxx5fp70fc9lq-elm-land-0.20.1/lib/node_modules/elm-land/node_modules/rollup/dist/native.js:59
		throw new Error(
		      ^

Error: Cannot find module @rollup/rollup-linux-x64-gnu. npm has a bug related to optional dependencies (https://github.com/npm/cli/issues/4828). Please try `npm i` again after removing both package-lock.json and node_modules directory.
    at requireWithFriendlyError (/nix/store/mbbvs7bg3n90gj0fva1zxx5fp70fc9lq-elm-land-0.20.1/lib/node_modules/elm-land/node_modules/rollup/dist/native.js:59:9)
    at Object.<anonymous> (/nix/store/mbbvs7bg3n90gj0fva1zxx5fp70fc9lq-elm-land-0.20.1/lib/node_modules/elm-land/node_modules/rollup/dist/native.js:68:76)
    ... 3 lines matching cause stack trace ...
    at Module._load (node:internal/modules/cjs/loader:1104:12)
    at cjsLoader (node:internal/modules/esm/translators:346:17)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:286:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:234:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:473:24) {
  [cause]: Error: Cannot find module '@rollup/rollup-linux-x64-gnu'
  Require stack:
  - /nix/store/mbbvs7bg3n90gj0fva1zxx5fp70fc9lq-elm-land-0.20.1/lib/node_modules/elm-land/node_modules/rollup/dist/native.js
      at Module._resolveFilename (node:internal/modules/cjs/loader:1225:15)
      at Module._load (node:internal/modules/cjs/loader:1051:27)
      at Module.require (node:internal/modules/cjs/loader:1311:19)
      at require (node:internal/modules/helpers:179:18)
      at requireWithFriendlyError (/nix/store/mbbvs7bg3n90gj0fva1zxx5fp70fc9lq-elm-land-0.20.1/lib/node_modules/elm-land/node_modules/rollup/dist/native.js:41:10)
      at Object.<anonymous> (/nix/store/mbbvs7bg3n90gj0fva1zxx5fp70fc9lq-elm-land-0.20.1/lib/node_modules/elm-land/node_modules/rollup/dist/native.js:68:76)
      at Module._compile (node:internal/modules/cjs/loader:1469:14)
      at Module._extensions..js (node:internal/modules/cjs/loader:1548:10)
      at Module.load (node:internal/modules/cjs/loader:1288:32)
      at Module._load (node:internal/modules/cjs/loader:1104:12) {
    code: 'MODULE_NOT_FOUND',
    requireStack: [
      '/nix/store/mbbvs7bg3n90gj0fva1zxx5fp70fc9lq-elm-land-0.20.1/lib/node_modules/elm-land/node_modules/rollup/dist/native.js'
    ]
  }
}

Node.js v20.17.0
```

**Comment by turboMaCk** [comment]
ah right. I also ran into this issue in my quick spike - that's where I gave up. Forgot to test it in this PR.

Anyway lets try fix these issues one by one now when this is merged. I also think we can start decoupling these problematic packages away from node2nix to buildNpmPackage.

File: pkgs/development/compilers/elm/packages/node/default.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (zupo) [reply pkgs/development/compilers/elm/packages/node/default.nix:22]: Tried again today during NixCon Hack Day: https://github.com/NixOS/nixpkgs/pull/351767

**Comment by domenkozar** [comment]
I tried using buildNpmPackage, but build from source is quite complicated as it needs to build parts using elm. Going to try to patch @rollup/wasm-node to be used in place of rollup.

**Comment by zupo** [comment]
I have gotten a bit further with `elm-land`, it now builds and runs on x86-linux and aarch64-darwin: https://github.com/NixOS/nixpkgs/pull/382339
